cpp-lite/include/mxnet-lite/ndarray.h

```diff
@@ -0,0 +1,284 @@
+                             context.GetDeviceId(), false, &handle), 0);
+    size_t size = shape.size() ? 1 : 0;
+    for (auto& i : shape) size *= i;
+    MXNDArraySyncCopyFromCPU(handle, data, size);
+    blob_ptr_ = std::make_shared<NDBlob>(handle);
+  }
+
+  void CopyFrom(const float *data, size_t size) {
+    CHECK_EQ(MXNDArraySyncCopyFromCPU(blob_ptr_->handle, data, size), 0);
+  }
+
+  void CopyTo(float *data, size_t size = 0) {
+     CHECK_EQ(MXNDArraySyncCopyToCPU(blob_ptr_->handle, data,
+                                     size > 0 ? size : GetSize()), 0);
+  }
+  /*!
+  * \brief return a new copy to this NDArray
+  * \param Context the new context of this NDArray
+  * \return the new copy
+  */
+  NDArray Copy(const Context &) const {
+    return NDArray();
+  }
+  /*!
+  * \brief Block until all the pending write operations with respect
+  *    to current NDArray are finished, and read can be performed.
+  */
+  void WaitToRead() const {
+    CHECK_EQ(MXNDArrayWaitToRead(blob_ptr_->handle), 0);
+  }
+  /*!
+  * \brief Block until all the pending read/write operations with respect
+  *    to current NDArray are finished, and write can be performed.
+  */
+  void WaitToWrite() const {
+    CHECK_EQ(MXNDArrayWaitToWrite(blob_ptr_->handle), 0);
+  }
+  /*!
+  * \return the size of current NDArray, a.k.a. the production of all shape dims
+  */
+  size_t GetSize() const {
+    size_t ret = 1;
+    for (auto &i : GetShape()) ret *= i;
+    return ret;
+  }
+  /*!
+  * \return the shape of current NDArray, in the form of mx_uint vector
+  */
+  std::vector<mx_uint> GetShape() const {
+    const mx_uint *out_pdata;
+    mx_uint out_dim;
+    CHECK_EQ(MXNDArrayGetShape(blob_ptr_->handle, &out_dim, &out_pdata), 0);
+    std::vector<mx_uint> ret;
+    for (mx_uint i = 0; i < out_dim; ++i) {
+      ret.push_back(out_pdata[i]);
+    }
+    return ret;
+  }
+  /*!
+  * \return the data type of current NDArray
+  */
+  int GetDType() const {
+    int ret;
+    CHECK_EQ(MXNDArrayGetDType(blob_ptr_->handle, &ret), 0);
+    return ret;
+  }
+  /*!
+  * \brief Get the pointer to data (IMPORTANT: The ndarray should not be in GPU)
+  * \return the data pointer to the current NDArray
+  */
+  const float *GetData() const {
+    if (GetDType() != 0) {
+      return NULL;
+    }
+    void *ret;
+    CHECK_EQ(MXNDArrayGetData(blob_ptr_->handle, &ret), 0);
+    return static_cast<float*>(ret);
+  }
+  /*!
+  * \return the context of NDArray
+  */
+  Context GetContext() const {
+    int out_dev_type;
+    int out_dev_id;
+    CHECK_EQ(MXNDArrayGetContext(blob_ptr_->handle, &out_dev_type,
+                                 &out_dev_id), 0);
+    return Context(static_cast<DeviceType>(out_dev_type), out_dev_id);
+  }
+  /*!
+  * \return the NDArrayHandle of the current NDArray
+  */
+  NDArrayHandle GetHandle() const {
+    return blob_ptr_->handle;
+  }
+
+ private:
+  /*!
+  * \brief struct to store NDArrayHandle
+  */
+  struct NDBlob {
```

Comment:
If this would be another private class and would contain most of the implementation of methods on NDArray that would be already an effective separation of the interface and implementation